### PR TITLE
Grammar apex

### PR DESCRIPTION
**Pull request by loqo71la**
Improve grammar rules to support DML statements

### CODE REVIEW

**Comment by loqo71la** [comment]
@marcocdlv @JimmyRojas please review our changes.

File: apex-checks/src/main/java/org/fundacionjala/enforce/sonarqube/apex/checks/DmlCheckInFor.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+import org.sonar.check.Rule;
+import org.sonar.squidbridge.annotations.ActivatedByDefault;
+import org.sonar.squidbridge.annotations.SqaleConstantRemediation;
+import org.sonar.squidbridge.annotations.SqaleSubCharacteristic;
+import org.sonar.squidbridge.checks.SquidCheck;
+
+import org.fundacionjala.enforce.sonarqube.apex.api.grammar.RuleKey;
+
+/**
+ * Check for a DML is not within a "for"
```
Comment (JimmyRojas) [inline apex-checks/src/main/java/org/fundacionjala/enforce/sonarqube/apex/checks/DmlCheckInFor.java]: "for"

"for loop"

File: apex-squid/src/main/java/org/fundacionjala/enforce/sonarqube/apex/api/ApexGrammar.java [hunk 2]
```diff
@@ -267,39 +346,89 @@ private static void typeClass(ApexGrammarBuilder grammarBuilder) {
...
+                LPAREN,
+                grammarBuilder.optional(PARAMETER_LIST),
+                RPAREN,
+                STATEMENT_BLOCK
         );
     }
 
     /**
-     * Creates rules for the header of a method.
+     * Creates rules to the last line of the method and the completion of the method.
```
Comment (JimmyRojas) [inline apex-squid/src/main/java/org/fundacionjala/enforce/sonarqube/apex/api/ApexGrammar.java]: Creates rules to the last line of the method and the completion of the method.

Creates rules for method's last line and completion

File: apex-squid/src/main/java/org/fundacionjala/enforce/sonarqube/apex/api/ApexGrammarBuilder.java [hunk 3]
```diff
@@ -50,11 +50,31 @@
     private static final String IDENTIFIER_PATTERN = "(_{0,2}[a-zA-Z][a-zA-Z0-9]*)+";
 
     /**
+     * Stores an numeric pattern.
```
Comment (JimmyRojas) [inline apex-squid/src/main/java/org/fundacionjala/enforce/sonarqube/apex/api/ApexGrammarBuilder.java]: Stores an numeric ...

Stores a numeric ...

File: apex-squid/src/main/java/org/fundacionjala/enforce/sonarqube/apex/api/ApexGrammarBuilder.java [hunk 4]
```diff
@@ -50,11 +50,31 @@
...
+     */
+    private static final String NUMERIC_PATTERN = "(0|[1-9]\\d*)";
+
+    /**
+     * Stores an CHARACTER pattern.
```
Comment (JimmyRojas) [inline apex-squid/src/main/java/org/fundacionjala/enforce/sonarqube/apex/api/ApexGrammarBuilder.java]: Stores an CHARACTER ...

Stores a CHARACTER ...

why uppercase?

File: apex-squid/src/main/java/org/fundacionjala/enforce/sonarqube/apex/api/ApexGrammarBuilder.java [hunk 5]
```diff
@@ -300,6 +392,26 @@ private boolean isIdentifier(TokenType token) {
     }
 
     /**
+     * Determines if a {@link TokenType} represents an numeric.
```
Comment (JimmyRojas) [inline apex-squid/src/main/java/org/fundacionjala/enforce/sonarqube/apex/api/ApexGrammarBuilder.java]: Determines if a {@link TokenType} represents an numeric.

Determines if a {@link TokenType} represents a numeric.

**Comment by JimmyRojas** [comment]
see some comments above

File: apex-checks/src/main/java/org/fundacionjala/enforce/sonarqube/apex/checks/DeprecatedMethodCheck.java [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+import org.sonar.squidbridge.annotations.ActivatedByDefault;
+import org.sonar.squidbridge.annotations.SqaleConstantRemediation;
+import org.sonar.squidbridge.annotations.SqaleSubCharacteristic;
+import org.sonar.squidbridge.checks.SquidCheck;
+
+import org.fundacionjala.enforce.sonarqube.apex.api.ApexKeyword;
+import org.fundacionjala.enforce.sonarqube.apex.api.grammar.RuleKey;
+
+/**
+ * Check that a deprecated method does not contain lines of code.
```
Comment (JimmyRojas) [inline apex-checks/src/main/java/org/fundacionjala/enforce/sonarqube/apex/checks/DeprecatedMethodCheck.java]: Checks

This class checks

File: apex-checks/src/main/java/org/fundacionjala/enforce/sonarqube/apex/checks/DeprecatedMethodCheck.java [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+     *
+     * @param astNode is the node to analyze.
+     * @return A value of true, if the node is expected.
+     */
+    private boolean validateToken(List<AstNode> astNode) {
+        return astNode.get(1).getTokenValue().equals(ApexKeyword.DEPRECATED.getValue());
+    }
+
+    /**
+     * Check if the node BLOCK_STATEMENT has content.
```
Comment (JimmyRojas) [inline apex-checks/src/main/java/org/fundacionjala/enforce/sonarqube/apex/checks/DeprecatedMethodCheck.java]: Checks

**Comment by JimmyRojas** [comment]
just one comment there, and it looks good to me

**Comment by loqo71la** [comment]
@JimmyRojas @marcocdlv please review our changes. we don't have more pending changes

File: apex-checks/src/main/java/org/fundacionjala/enforce/sonarqube/apex/checks/DeprecatedMethodCheck.java [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+        return astNode.get(1).getTokenValue().equals(ApexKeyword.DEPRECATED.getValue());
+    }
+
+    /**
+     * Checks if the node BLOCK_STATEMENT has content.
+     *
+     * @param astNode is the node to analyze.
+     * @return If the node has content, returns the same node, null if has not.
+     */
+    private AstNode validateStatementBlock(List<AstNode> astNode) {
```
Comment (JimmyRojas) [inline apex-checks/src/main/java/org/fundacionjala/enforce/sonarqube/apex/checks/DeprecatedMethodCheck.java]: too much work to verify if the method has body inside, please review and improve the validation

File: apex-checks/src/main/java/org/fundacionjala/enforce/sonarqube/apex/checks/DmlInForCheck.java [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+
+    /**
+     * It is responsible for verifying whether the rule is met in the rule base.
+     * In the event that the rule is not correct, create message error.
+     *
+     * @param astNode It is the node that stores all the rules.
+     */
+    @Override
+    public void visitNode(AstNode astNode) {
+        if (astNode.hasDescendant(RuleKey.DML_STATEMENT)) {
```
Comment (JimmyRojas) [inline apex-checks/src/main/java/org/fundacionjala/enforce/sonarqube/apex/checks/DmlInForCheck.java]: this validation is happening several time, can we try to abstract?

File: apex-checks/src/main/java/org/fundacionjala/enforce/sonarqube/apex/checks/DmlInWhileCheck.java [hunk 10]
```diff
@@ -36,65 +36,49 @@
...
-     *
-     * @param node used to compare.
-     */
-    private void checkNode(AstNode node) {
-        if (firstNode == null) {
-            firstNode = node;
-        } else if (!firstNode.getTokenValue().equals(node.getTokenValue())) {
-            getContext().createLineViolation(this,
-                    String.format(MESSAGE, node.getToken().getValue(), firstNode.getToken().getValue()), node);
+    public void visitNode(AstNode astNode) {
```
Comment (JimmyRojas) [inline apex-checks/src/main/java/org/fundacionjala/enforce/sonarqube/apex/checks/DmlInWhileCheck.java]: this validation is happening several time, can we try to abstract?

File: apex-checks/src/main/java/org/fundacionjala/enforce/sonarqube/apex/checks/DmlInConstructorCheck.java [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+    }
+
+    /**
+     * It is responsible for verifying whether the rule is met in the rule base.
+     * In the event that the rule is not correct, create message error.
+     *
+     * @param astNode It is the node that stores all the rules.
+     */
+    @Override
+    public void visitNode(AstNode astNode) {
```
Comment (JimmyRojas) [inline apex-checks/src/main/java/org/fundacionjala/enforce/sonarqube/apex/checks/DmlInConstructorCheck.java]: this validation is happening several time, can we try to abstract?

**Comment by JimmyRojas** [comment]
please see a few comments above

**Comment by RockFercho** [comment]
@JimmyRojas @marcocdlv please, review my changes.

**Comment by JimmyRojas** [comment]
now it looks better, no more comments from my side.. :)

File: apex-checks/src/main/java/org/fundacionjala/enforce/sonarqube/apex/checks/AssertMethodCheck.java [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+import org.fundacionjala.enforce.sonarqube.apex.api.ApexKeyword;
+import org.fundacionjala.enforce.sonarqube.apex.api.grammar.RuleKey;
+
+/**
+ * Verifies if a test method contains invalid asserts.
+ */
+@Rule(
+        key = AssertMethodCheck.CHECK_KEY,
+        priority = Priority.MINOR,
+        name = "Invalid asserts in test methods",
```
Comment (ziur) [inline apex-checks/src/main/java/org/fundacionjala/enforce/sonarqube/apex/checks/AssertMethodCheck.java:48]: It is ok to have asserts in test methods, what is wrong is bad practices like assert(true), please change the name

File: apex-checks/src/main/java/org/fundacionjala/enforce/sonarqube/apex/checks/AssertMethodCheck.java [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+     * In the event that the rule is not correct, create message error.
+     *
+     * @param astNode It is the node that stores all the rules.
+     */
+    @Override
+    public void visitNode(AstNode astNode) {
+        if (!isTest(astNode)) {
+            return;
+        }
+        List<AstNode> expressions = astNode.getDescendants(RuleKey.ARGUMENTS);
```
Comment (ziur) [inline apex-checks/src/main/java/org/fundacionjala/enforce/sonarqube/apex/checks/AssertMethodCheck.java:93]: I think we can improve this part, we don't need to look for all ARGUMENTS nodes, just nodes related to assert, in this way we can reduce the amount of nodes to evaluate

File: apex-checks/src/main/java/org/fundacionjala/enforce/sonarqube/apex/checks/DeprecatedMethodCheck.java [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+
+import org.fundacionjala.enforce.sonarqube.apex.api.grammar.RuleKey;
+
+/**
+ * This class checks that a deprecated method does not contain lines of code.
+ */
+@Rule(
+        key = DeprecatedMethodCheck.CHECK_KEY,
+        priority = Priority.INFO,
+        name = "Verification methods deprecated",
```
Comment (ziur) [inline apex-checks/src/main/java/org/fundacionjala/enforce/sonarqube/apex/checks/DeprecatedMethodCheck.java:42]: We need to start considering use a resource bundle for messages, and use enumerations for those messages.

File: apex-checks/src/main/java/org/fundacionjala/enforce/sonarqube/apex/checks/DmlInWhileCheck.java [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+
+import org.fundacionjala.enforce.sonarqube.apex.api.grammar.RuleKey;
+
+/**
+ * Check for a DML is not within a "while"
+ */
+@Rule(
+        key = DmlInWhileCheck.CHECK_KEY,
+        priority = Priority.CRITICAL,
+        name = "You can not be a DML statement in a 'while'",
```
Comment (ziur) [inline apex-checks/src/main/java/org/fundacionjala/enforce/sonarqube/apex/checks/DmlInWhileCheck.java:41]: Need to improve the message

**Pull request merged**